### PR TITLE
Adding a course from find for a signed in candidate

### PR DESCRIPTION
**Pull request by davidgisbey**
## Context
**https://trello.com/c/c4ZYOuaI/1010-add-you-selected-a-course-page-into-find-to-apply-journey will be affected by this work**

This PR covers stretch goal 1 of the Trello card. This is to fix adding a course from find for candidates who are already signed in.
 
## Changes proposed in this pull request

- Rework the start page controller to call the ExistingCandidateAuthentication service if the user is signed in and course params are present

## Guidance to review

I think the naming of the ExistingCandidateAuthentication does not really describe what it actually does,  but I can't come up with anything that seems to describe it well.

# Link to Trello card

https://trello.com/c/71lG72Y6/

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/controllers/candidate_interface/start_page_controller.rb [hunk 1]
```diff
@@ -26,5 +26,38 @@ def eligibility_params
...
         .transform_values(&:strip)
     end
+
+  private
+
+    def redirect_signed_in_candidate
+      if current_candidate.present?
+        add_course_from_find_id_to_candidate if params[:providerCode].present?
+
+        service = ExistingCandidateAuthentication.new(candidate: current_candidate)
```
Comment (davidgisbey) [inline app/controllers/candidate_interface/start_page_controller.rb]: Would love any thoughts on what this service should be renamed to. The current naming doesn't really describe what is it doing

File: spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb [hunk 2]
```diff
@@ -0,0 +1,173 @@
+require 'rails_helper'
+
+RSpec.describe 'An existing candidate arriving from Find with a course and provider code' do
+  include CourseOptionHelpers
+  scenario 'candidate is not signed in and retains their course selection through the sign up process' do
```
Comment (fofr) [inline spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb]:
```suggestion
  scenario 'candidate is signed in' do
```

Comment (davidgisbey) [reply spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb]: thanks

File: spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    and_i_am_signed_in
+    and_i_have_less_than_3_application_options
+
+    when_i_arrive_at_the_apply_from_find_page_with_course_params_with_multiple_sites
+    and_i_click_apply_on_apply
+    then_i_should_see_the_course_choices_site_page
+    and_i_see_the_form_to_pick_a_location
+
+    given_the_course_i_selected_has_multiple_sites
+    and_i_am_an_existing_candidate_on_apply
```
Comment (stevehook) [inline spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb]: Creating a new candidate in the middle of the test confused me a bit. But I think we should probably leave as is - we don't want to break this up and multiply the number of tests we have. What do you think?

File: spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    create(:course_option, site: @site, course: @course, vacancy_status: 'B')
+  end
+
+  def and_i_am_an_existing_candidate_on_apply
+    @email = "#{SecureRandom.hex}@example.com"
+    @candidate = create(:candidate, email_address: @email)
+  end
+
+  def and_i_have_less_than_3_application_options
+    application_form = create(:application_form, candidate: @candidate)
```
Comment (stevehook) [inline spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb]: Does creating more than one application form for the same candidate cause problems? (We assume elsewhere that the `current_application` is `candidate.application_forms.first`

**Comment by tvararu** [comment]
Works great in my testing on the review app 💯

File: spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb (hunk 3 above)
Comment (davidgisbey) [reply spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb]: Yea I get that and it definitely could be split over more test files. I agree though that the overhead of creating the extra files and duplicating the majority of the methods probably isn't worth it.

File: spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb (hunk 4 above)
Comment (davidgisbey) [reply spec/system/candidate_interface/candidate_signed_in_user_with_course_params_spec.rb]: Good spot. I think following the pattern of creating a new candidate is probably preferable considering you already have to parse it once anyway.

**Pull request merged**